### PR TITLE
Do not suggest patterns in expression trees

### PR DESCRIPTION
**Pull request by alrz**
Address two issues mentioned in https://github.com/dotnet/roslyn/pull/43596#issuecomment-645606377

/cc @CyrusNajmabadi

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs [hunk 1]
```diff
@@ -82,6 +85,7 @@ class C
         [InlineData("o != null")]
         [InlineData("!(o is null)")]
         [InlineData("o is int ii || o is long jj")]
+        [InlineData("i == default || i > default(int)")]
```
Comment (alrz) [inline src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs:88]: The operation APIs cannot get a type for `default`, hence no constants.  Perhaps I misread sth last time..

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs:88]: > The operation APIs cannot get a type for default, hence no constants. Perhaps I misread sth last time..

That seems super weird.  all IOps shoudl expose their type.   @333fred ?

File: src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -38,23 +38,29 @@ protected override void InitializeWorker(AnalysisContext context)
...
                 return;
 
+            var cancellationToken = context.CancellationToken;
             var styleOption = context.Options.GetOption(CSharpCodeStyleOptions.PreferPatternMatching, syntaxTree, cancellationToken);
             if (!styleOption.Value)
                 return;
 
-            var operation = context.SemanticModel.GetOperation(expression, cancellationToken);
+            var semanticModel = context.SemanticModel;
+            var expressionTypeOpt = context.Compilation!.GetTypeByMetadataName("System.Linq.Expressions.Expression`1");
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs:59]:
```suggestion
            var expressionTypeOpt = semanticModel.Compilation.GetTypeByMetadataName("System.Linq.Expressions.Expression`1");
```

File: src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs (hunk 1 above)
Comment (333fred) [reply src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs:88]: A `default` literal has no type. It should have a conversion on top of it with the converted type.

Comment (alrz) [reply src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs:88]: >A default literal has no type. It should have a conversion on top of it with the converted type.

Ah that makes sense. But we won't have `ConstantValue` still, right?

File: src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -38,23 +38,29 @@ protected override void InitializeWorker(AnalysisContext context)
         private void AnalyzeNode(SyntaxNodeAnalysisContext context)
         {
             var expression = (ExpressionSyntax)context.Node;
-
+            if (expression.GetDiagnostics().Any(diagnostic => diagnostic.Severity == DiagnosticSeverity.Error))
```
Comment (alrz) [inline src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs:41]: so my test case had an error on `i` (instance field access in instance field initializer), I'm surprised that this didn't catch it..

File: src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs [hunk 4]
```diff
@@ -88,6 +94,7 @@ public async Task TestMissingOnExpression(string expression)
             await TestAllMissingOnExpressionAsync(expression);
         }
 
+        [InlineData("i == default || i > default(int)", "i is default(int) or > (default(int))")]
```
Comment (alrz) [inline src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs:97]: simplifier doesn't handle relational patterns. I didn't go through that - probably best to be done separately. /cc @CyrusNajmabadi

File: src/Features/CSharp/Portable/UsePatternCombinators/CSharpUsePatternCombinatorsCodeFixProvider.cs [hunk 5]
```diff
@@ -89,7 +89,7 @@ private static PatternSyntax AsPatternSyntax(AnalyzedPattern pattern)
                     Token(p.Token.LeadingTrivia, p.IsDisjunctive ? SyntaxKind.OrKeyword : SyntaxKind.AndKeyword,
                         TriviaList(p.Token.GetAllTrailingTrivia())),
                     AsPatternSyntax(p.Right).Parenthesize()),
-                Constant p => ConstantPattern(p.ExpressionSyntax.Parenthesize()),
+                Constant p => ConstantPattern(AsExpressionSyntax(p)),
                 Source p => p.PatternSyntax,
                 Type p => TypePattern(p.TypeSyntax),
                 Relational p => RelationalPattern(Token(MapToSyntaxKind(p.OperatorKind)), p.Value.Parenthesize()),
```
Comment (alrz) [inline src/Features/CSharp/Portable/UsePatternCombinators/CSharpUsePatternCombinatorsCodeFixProvider.cs:95]: I didn't do the same rewrite for relational pattern - I believe it's a compiler bug. Filed https://github.com/dotnet/roslyn/issues/45292

**Pull request merged**